distribution/cachedir/manager.go

```diff
@@ -0,0 +1,190 @@
+}
+
+type manager struct {
+	cacheRoot string
+	mu        sync.Mutex
+	inUse     map[HashedKey]struct{}
+}
+
+// NewManager creates a new cachedir manager
+func NewManager(path string) Manager {
+	return &manager{
+		cacheRoot: path,
+		inUse:     map[HashedKey]struct{}{},
+	}
+}
+
+// HashedKey is a sha256 hash of a cache key which is used to allow using a
+// fixed length but sufficiently unique directory name while allowing "keys" to
+// be of arbitrary length
+type HashedKey string
+
+// hashKey hashes the key to create the directory name because otherwise tche path
+// would be too long. This hash needs to be secure or else multiple downloads could
+// share the same cache directory and cause unexpected and potentially malicious behaviour.
+func hashKey(key string) HashedKey {
+	hash := sha256.New()
+	hash.Write([]byte(key))
+	encoded := make([]byte, hash.Size()*2)
+	hex.Encode(encoded, hash.Sum(nil))
+	return HashedKey(encoded)
+}
+
+func (c *manager) cachePathForKeyHash(keyHash HashedKey) string {
+	return filepath.Join(c.cacheRoot, string(keyHash))
+}
+
+func dirSize(path string) (uint64, error) {
+	var size uint64
+	err := filepath.Walk(path, func(_ string, info os.FileInfo, err error) error {
+		if info != nil && !info.IsDir() {
+			size += uint64(info.Size())
+		}
+		return err
+	})
+	return size, err
+}
+
+func (c *manager) sizeEstimateByKeyHash(keyHash HashedKey) uint64 {
+	dir := c.cachePathForKeyHash(keyHash)
+	size, err := dirSize(dir)
+	if err != nil {
+		logrus.WithError(err).WithField("cachedir", dir).Warnf("failed to determine size of directory")
+		return 0
+	}
+	return size
+}
+
+func (c *manager) cleanUpByKeyHash(keyHash HashedKey) error {
+	return os.RemoveAll(c.cachePathForKeyHash(keyHash))
+}
+
+func (c *manager) GetDir(key string) string {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	hashedKey := hashKey(key)
+	logrus.WithField("key", key).WithField("hashedKey", hashedKey).Debugf("cachedir manager get")
+	if _, ok := c.inUse[hashedKey]; ok {
+		panic(fmt.Errorf("tried to get a cachedir key (%s) twice without releasing/deleting", key))
+	}
+	c.inUse[hashedKey] = struct{}{}
+
+	return c.cachePathForKeyHash(hashedKey)
+}
+
+func (c *manager) ReleaseDir(key string) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	hashedKey := hashKey(key)
+	logrus.WithField("key", key).WithField("hashedKey", hashedKey).Debugf("cachedir manager release")
+	if _, ok := c.inUse[hashedKey]; !ok {
+		panic(fmt.Errorf("tried to release a cachedir key (%s) without getting it first", key))
+	}
+	delete(c.inUse, hashedKey)
+}
+
+func (c *manager) DeleteDir(key string) error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+	hashedKey := hashKey(key)
+	logrus.WithField("key", key).WithField("hashedKey", hashedKey).Debugf("cachedir manager delete")
+	if _, ok := c.inUse[hashedKey]; !ok {
+		panic(fmt.Errorf("tried to delete a cachedir key (%s) without getting it first", key))
+	}
+	delete(c.inUse, hashedKey)
+	return c.cleanUpByKeyHash(hashedKey)
```

Comment:
I think it would be okay to release `c.mu` before doing this I/O, so we don't block other calls to `manager`.